src/aof.c

```diff
@@ -40,189 +40,692 @@
+ * line, followed by history type AOFs and finally is the INCR AOFs.
+ */
+#define AOF_INFO_FORMAT_AND_CAT(buf, info)                    \
+    sdscatprintf(buf, "%s %s %s %lld %s %c\n",                \
+                 AOF_MANIFEST_KEY_FILE_NAME, info->file_name, \
+                 AOF_MANIFEST_KEY_FILE_SEQ, info->file_seq,   \
+                 AOF_MANIFEST_KEY_FILE_TYPE, info->file_type)
+
+sds getAofManifestAsString(aofManifest *am) {
+    serverAssert(am != NULL);
+
+    sds buf = sdsempty();
     listNode *ln;
     listIter li;
-    unsigned long size = 0;
 
-    listRewind(server.aof_rewrite_buf_blocks,&li);
-    while((ln = listNext(&li))) {
-        aofrwblock *block = listNodeValue(ln);
-        size += block->used;
+    /* 1. Add BASE File information, it is always at the beginning 
+     * of the manifest file. */
+    if (am->base_aof_info) {
+        buf = AOF_INFO_FORMAT_AND_CAT(buf, am->base_aof_info);
+    }
+    
+    /* 2. Add HISTORY type AOF information. */
+    listRewind(am->history_aof_list, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        aofInfo *ai = (aofInfo*)ln->value;
+        buf = AOF_INFO_FORMAT_AND_CAT(buf, ai);
     }
-    return size;
-}
-
-/* This function is different from aofRewriteBufferSize, to get memory usage,
- * we should also count all other fields(except 'buf') of aofrwblock and the
- * last block's free size. */
-unsigned long aofRewriteBufferMemoryUsage(void) {
-    unsigned long size = aofRewriteBufferSize();
 
-    listNode *ln = listLast(server.aof_rewrite_buf_blocks);
-    if (ln != NULL) {
-        aofrwblock *block = listNodeValue(ln);
-        size += block->free;
-        size += (offsetof(aofrwblock,buf) *
-                 listLength(server.aof_rewrite_buf_blocks));
+    /* 3. Add INCR type AOF information. */
+    listRewind(am->incr_aof_list, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        aofInfo *ai = (aofInfo*)ln->value;
+        buf = AOF_INFO_FORMAT_AND_CAT(buf, ai);
     }
-    return size;
+
+    return buf;
 }
 
-/* Event handler used to send data to the child process doing the AOF
- * rewrite. We send pieces of our AOF differences buffer so that the final
- * write when the child finishes the rewrite will be small. */
-void aofChildWriteDiffData(aeEventLoop *el, int fd, void *privdata, int mask) {
-    listNode *ln;
-    aofrwblock *block;
-    ssize_t nwritten;
-    mstime_t latency;
-    UNUSED(el);
-    UNUSED(fd);
-    UNUSED(privdata);
-    UNUSED(mask);
+/*  Load the manifest information from the disk to `server.aof_manifest` 
+ *  when the redis server start.
+ *  
+ *  During the loading process, we will conduct strict error checking. 
+ *  Once there are file opening error, format error, etc., we will 
+ *  directly exit the redis process.
+ * 
+ *  Note: We will ignore the "doesn't exist" error, because this will 
+ *  happen when we upgrade from an old version redis.
+ */
+void aofLoadManifestFromDisk(void) {
```

Comment:
1. I think there is no need to pass the aofManifest parameter here. As you can see, in all the functions I implemented, some passed the aofManifest parameter and some did not. The principle is that if it is a public API, such as being used in server.c, then it is the top-level API, and then it can directly access `server.aof_manifest`. Other functions with aofManifest parameters , They are mainly used internally (aof.c), and may handle two cases of `server.aof_manifest` and `temp_manifest`, so they need a variable parameter.

There are similar top-level APIs (declared in server.h):
```
void aofLoadManifestFromDisk(void);
void aofOpenIfNeededOnServerStart(void);
int aofDelHistoryFiles(void);
int aofRewriteLimited(void);
```

2. I don't think a return value is needed here. The reason is that manifest is a very important file (even more important than `redis.conf`), so once an error occurs, it is safest to exit directly and print the error message.

The following is the function declaration for loading redis.conf:
```
void loadServerConfig(char *filename, char config_from_stdin, char *options);
```